### ISSUE TITLE
Constant crashing in 2.0 after several minutes only with Alliance AI enabled

### ISSUE THREAD

**Issue by tdauth** [tdauth/wowr#615]
Fix the AI scripts or building pathings.
Maybe the new portals cause the crash for the Alliance AI.

**Comment by tdauth** [comment]
After defeating the alliance it crashed after like 30 minutes.

**Comment by tdauth** [comment]
Find a tool which can analyze replays for the final actions in the replay. The action before the crash might have caused it.

**Comment by tdauth** [comment]
https://www.hiveworkshop.com/threads/replay-analysis-tool-which-lists-all-final-actions-from-all-computer-players.338509/

**Comment by tdauth** [comment]
After around 30 minutes, purple was attacking King Deepbeard. Maybe it is some timer after 30 minutes?

**Comment by tdauth** [comment]
Maybe it is the Statue of Azshara. When I revealed the whole map it was casting constantly to The Burning Legion.

**Comment by tdauth** [comment]
Not the special buildings. I tested both.

**Comment by tdauth** [comment]
Got the crash with both special players enabled after 24:40 minutes with warpten cheat.
Maybe it is some attack wave. Other player wrote on Discord after 20 minutes.

**Comment by tdauth** [comment]
Murgol Reaver hat AImove commands shortly before the crash.

**Comment by tdauth** [comment]
public constant [integer](http://wiki.thehelper.net/wc3/jass/common.j/integer) AImove=851988

**Comment by tdauth** [comment]
ACCESS_VIOLATION DBG-OPTIONS<FunctionsOnly SingleLine> DBG-ADDR<00007FF705856873>("Warcraft III.exe") <- DBG-ADDR<00007FF706C96F52>("Warcraft III.exe") <- DBG-ADDR<00007FF705C6B15B>("Warcraft III.exe")  DBG-OPTIONS<>

**Comment by tdauth** [comment]
<Mem.PhysicalUsed>3093 MB
<Mem.VirtualUsed>8578 MB
<Mem.Committed>3622 MB
<Mem.PhysicalAvailable>3442 MB
<Mem.VirtualAvailable>134209150 MB
<Mem.PageAvailable>3880 MB

**Comment by tdauth** [comment]
85197 Morgul Slave
852660 same

**Comment by tdauth** [comment]
The Alliance unloadall Zeppelin.

**Comment by tdauth** [comment]
851974 Blood Elf Male Citizen

**Comment by tdauth** [comment]
Finger of Death Eredar Sorcerer on Goldmine and after that and 851974 on eredar sorcerer.

**Comment by tdauth** [comment]
Maybe:
- SetRandomPaths

or the final attack wave of The Alliance:

```
		call CampaignAttackerEx( 4, 4, 4, BLOOD_ELF_WAGON )
		call CampaignAttackerEx( 4, 4, 4, BLOOD_ELF_BALLISTA )
		call CampaignAttackerEx( 4, 4, 4, BLOOD_ELF_CAGE )
```

but Blood Elf AI had that before. Maybe just disable random paths.
Maybe some timer after 20-30 minutes.

**Comment by tdauth** [comment]
Maybe we can read the core dump on windows with something like the gdb:
https://stackoverflow.com/questions/8305866/how-do-i-analyze-a-programs-core-dump-file-with-gdb-when-it-has-command-line-pa
https://www.digitalcitizen.life/view-contents-dump-file/

**Comment by tdauth** [comment]
Check all AI related changes since 1.9.9:
- Reduce the number of air units and increase the number of ground units for Demon/TBL AIs.
-  Add new building type Research Tent with all the important researches from tier 3 buildings for all races.
- Reduce Demon Housing model scale to make build area visible.
- Let Naga AI build special building.
- Reduce attacking air units from The Alliance.
- Fix respawning The Alliance AI at the correct start location.
- Add fixed Computer player Gaia who provides some units on the map which can be rescued.
- The Alliance and Blood Elf AI build special building Magic Vault.
 
Maybe it really is the Research Tent. I used the Special Buildings without crashes but when AI uses it it might be different!

**Comment by tdauth** [comment]
Random Events and Alliance with Burning Legion did already exist. The portals and Water Portals are new. The Alliance can move to different locations now.

Here is the complete 2.0 ChangeLog:
- Fix number of Adventurers in the map description.
- Add cheat "-savegui".
- Add more unit, building, item and research  types for savecodes.
- Fix defense type for Turtle Shell item.
- Increase death time of creep/boss heroes to avoid their permanent removal from the game.
- Disease Cloud of Corrupted Ancient Protector does not require a research anymore.
- Recreate Furbolg Tribal Center on AI destruction.
- Do not remove Sargeras for non-Boss players on death.
- Set mana cost for all levels of Mirror Image to 80 which was 125 from level 4.
- Fix effect of item Amulet of Martial Arts and remove Unique from its tooltip.
- Add new boss Spider Crab Behemoth.
- Rename Sea Giant boss to King Deepbeard.
- Reveal certain neutral buildings to Freelancers.
- Show hero journey messages with colors, sound and in correct order when leveling up.
- Increase Archimonde's hero level to 150 and the level of his Dreadlords to level 60.
- Disable hero level up for Mirror Image of bosses because of Evolution.
- Move town halls of The Alliance next to the goldmines.
- Add line break to the defeat message.
- Translate victory message into English.
- Add chat command "-wrapup" which converts all buildings into tiny items.
- More general approach nor not picking up items which do not belong to your races.
- Add portal to and from Alliance island.
- Change color of portal to Archimonde to the The Burning Legion player color.
- Add portal to and from Boss Fight Island.
- Add chat command "-passive" for a passive mode against enemy players.
- Engineer Ships can build Haunted Water Gold Mines.
- Engineer Ships can place Goblin Water Mines.
- Increase stock number and decrease stock delay for Engineer Ships.
- Haunted Water Gold Mines can train ships and Sky Barges.
- Add research Improved Air Transport to the Power Generator.
- Add Mercenary Camp from the Future to the Boss Fight Island.
- Increase hitpoints and mana bonus from 50 % to 60 % for Evolution.
- Increase hero ability levels to 100.
- - Show the percentage in the tooltip of Brilliance Aura.
- Increase hero level of Archimonde to 500 which is the maximum hero level skill points can be spent.
- Remove irritating boss marker at mages creep spot in Lordaeron.
- Add new boss Mathog.
- Allow deactivating Divine Shield.
- Reduce cooldown and duration of Divine Shield.
- Add new boss Gul'dan.
- Simplify triggers to drop legendary artifacts on hero deaths.
- Add JASS function AssignAllUnitsToCurrentGroup to assign all units at once to a creep respawn group.
- Add global unit group variable udg_CurrentRespawnGroup to add all respawning creeps to it.
- Replace useless Embassy of Theramore with an Arcane Academy of Theramore selling mana-related items.
- Remove useless Tavern of Theramore.
- Add new island next to Maelstrom with new boss Sea Witch.
- Play Naga or Blood Elf Theme if you choose the corresponding races.
- Reduce the number of air units and increase the number of ground units for Demon/TBL AIs.
- Add new boss Avatar of Sargeras with legendary item which can convert units to different races.
- Fix worker counter in stats when using Charm.
- Add more unit and building types for savecodes.
- New race Goblins.
- Add save code for researches and chat command "-loadr XXX" to load it.
- Allow voting for Only Freelancers for AI in the beginning.
- Generate dragon units and good items savecodes with cheats.
- Replace Tier 1 - 3 equivalent texts with "Tier 1", "Tier 2" and "Tier 3" instead of listing all town halls.
- New island Kezan with boss Rumblefitz.
- Allow disabling AI The Burning Legion and The Alliance via votes in the beginning.
- Add chat command "-racerepick" to repick your first race as Warlord.
- Add Hall of Fame floating text to hero selection.
- Add building type Research Tent for all races with important researches to save icon space in tier 3 buildings.
- Fix hotkeys for bonus hero Medivh.
- Add chat command "-savec" to generate savecodes into files without printing information for a long time.
- Remove weather effect from hidden island.
- Add learn hotkey to ability Summon Sea Elemental of bonus hero Admiral Proudmoore.
- Remove message "Upgraded Evolution.".
- Reduce Demon Housing model scale to make build area visible.
- Let Naga AI build special building.
- Add more Hidden Bases.
- Add chat command "-saveautoon/off" and auto save the hero every 10 minutes.
- Adapt Warlord and Freelancer descriptions.
- Add portals to different boss islands on the sea.
- Always summon Sunken Ruin mercenaries for Freelancers when the regions are unknown.
- Fix respawning The Alliance AI at the correct start location.
- Add chat command "-help/-h" to list all chat commands.
- Becoming a demigod requires a Nether Dragon.
- Add chat command "-pingdragons" to ping all Dragon Roosts.
- Power Generator causes damage in area of boss Gul'dan.
- Add secret combination to enter Gul'dan's dungeon.
- Add treasure with trap to Gul'dan's dungeon.
- Fix hero revival for Furbolg AI.
- Support more heroes for Freelancer AI.
- Add leaderboards quest log entry from the original Reborn map.
- Invisibility and Berserk have 100 levels now and are set to the hero's level up to level 100.
- Both Pit Lords in Outland carry an Orb of Fire now to attack air units.
- Initial Towers of the Arcane Academy of Theramore allow changing hero abilities.
- Add fixed Computer player Gaia who provides some units on the map which can be rescued.
- Add research Improved Navy for all races.
- Human, Night Elf and Undead ships are affected by Evolution now.
- Added shipyards to The Alliance for initial navy AI support.
- Fix soundset of Nether Dragon boss.
- Use the hero proper name of bosses in the message for all players when defeating them.
- Add build blockers to the Outland portals area.
- The Alliance and Blood Elf AI build special building Magic Vault.
- Reset hero level researchs on full repick.
- Try to fix alliance change reaction by AI players by swapping player parameters.
- Add Water Waygates.
- Allow Engineer Ships to build Portals on water.
- Do not allow to load save objects which do not belong to your race.
- Decrease the save code IDs by splitting them up into units, buildings, items and researches.
- Increase maximum level of research Evolution from 100 to 1000.
- Do not increase the area of effect for War Stomp with each level to avoid lags.
- Disable AI special players by default to avoid crashes.
- Move creep dragons further away from the Orgrimmar start location.
- Add chat command "-pingkeys" to ping all keys which are not in black mask for a player.
- Replace Spell Immunity of bonus hero Archimonde with Invisibility.
- Add research Improved Creep Hunter which increases the maximum creep level of certain unit abilities.
- Fill savecodes with fixed number with remaining zeroes to always have a fixed length since we do not store the number.
- Fix cooldown for savecode researches.
- The chat commands "-save" and "-savegui" open a custom GUI in singleplayer only which allows managing savecodes.
- Do not load research levels from savecodes if your current ones are higher.
- Do not load savecodes if they are empty.
- Store Improved Navy level in the hero savecode.
- Fix an issue of loading not existing hero level 2 from the savecode.
- The savecode stores if your hero has become a Demigod or not.
- Add two more creep unit types (one healing) to the Neutral Mercenary Camp of Freelancers.
- Translate abilities of Tinker back into English.
- Allow saving fragment items required to become a Demigod.
- Reorder columns in the stats multiboard.
- Give Kil'Jaeden boss Banish ability.
- AutoSkillHero and simplify Boss Island boss creation.
- One single cheat "-levelX".
- New cheat "-cheats" to list all cheats.
- Ping Water Goldmines with "-pinggoldmines" chat command.
- Make Sargeras boss much stronger by giving him some powerful unit abilities.
- Do not allow to pick Sargeras boss as second hero.
- Add item Feed made by the Research Tent which increases the hero level of all bosses.
- Add third hero at hero level 100 with chat command "-thirdrepick".
- Add cheats "-orderon/off" to debug orders by all players to find crashes.
- Attribute Bonus has 1000 levels now.
- Sorcerer and Runeforger craft items with 4 charges instead of 3 now.
- Allow enabling AI for yourself at hero level 200.
- Clans.
- Reset abilities of mercenary and creep Dark Wizard to default.
- Give Maiev a Shadow Orb to attack air units.
- Add Demon Quests 3 and 4.
- Lower the priorities for Research Tent and Power Generator to avoid AI crashes.
- Bounties.
- Store item names, descriptions, tooltips and extended tooltips in backpacks to avoid losing dungeon key information.
- Add Goblin Bank.
- Enable Warcraft's Item Stacking.

**Comment by tdauth** [comment]
Not the Water Portals! I removed them in the beginning: Still a crash!

**Comment by tdauth** [comment]
Maybe picking a second/third hero. Maybe the AI players think they can pick another hero and it crashes?

**Comment by tdauth** [comment]
Disabling research tent did not help. Maybe some portals have the wrong target rect and lead to endless paths?

**Comment by tdauth** [comment]
The portals are all connected. What about reviving dead AI heroes in Altars?

**Comment by tdauth** [comment]
The Goblin Town Hall was chosen as Altar to revive a hero!

**Comment by tdauth** [comment]
Item Crafting for AI was enabled for The Alliance.

**Comment by tdauth** [comment]
Exclude Gaia everywhere!

**Comment by tdauth** [comment]
No Crash for 40 minutes without Burning Legion and The Alliance.

See https://www.hiveworkshop.com/threads/list-of-warcraft-iii-crashes.194706/
Do we use the Orb ability for Attack Ground?
Forked Lightning with 0 as its Stats - Area of Effect will freeze the game. for the Naga hero?

**Comment by tdauth** [comment]
Crashes again without Burning Legion etc. Maybe some raw codes are wrong (Research Tent, Research Tent upgrades, Freelancer AI!!!).

**Comment by tdauth** [comment]
Changes in the Freelancer AI script (only hero abilities not hero ids):
```

    constant integer TRANSMUTE                                       = 'ANtm'
    constant integer HEALING_SPRAY                                   = 'ANhs'
    constant integer ACID_BOMB                                       = 'ANab'
    constant integer CHEMICAL_RAGE                                   = 'ANcr'

    constant integer TINKER                                          = 'Ntin'

    constant integer ROBO_GOBLIN                                     = 'ANrg'
    constant integer CLUSTER_ROCKETS                                 = 'ANcs'
    constant integer ENGINEERING                                     = 'ANeg'
    constant integer POCKET_FACTORY                                  = 'ANsy'

    constant integer BREWMASTER                                      = 'Npbm'

    constant integer STORM_EARTH_AND_FIRE                            = 'ANef'
    constant integer BREATH_OF_FIRE                                  = 'ANbf'
    constant integer DRUNKEN_HAZE                                    = 'ANdh'
    constant integer DRUNKEN_BRAWLER                                 = 'ANdb'

    constant integer FIRELORD                                        = 'Nfir'

    constant integer VOLCANO                                         = 'ANvc'
    constant integer INCINERATE                                      = 'ANic'
    constant integer SOUL_BURN                                       = 'ANso'
    constant integer SUMMON_LAVA_SPAWN                               = 'ANlm'
```

skilling:

```
    elseif (curHero == ALCHEMIST) then
        if (skillUlti) then
            return TRANSMUTE
        elseif (mod == 0) then
            return HEALING_SPRAY
        elseif (mod == 1) then
            return ACID_BOMB
        elseif (mod == 2) then
            return CHEMICAL_RAGE
        endif
    elseif (curHero == TINKER) then
        if (skillUlti) then
            return ROBO_GOBLIN
        elseif (mod == 0) then
            return CLUSTER_ROCKETS
        elseif (mod == 1) then
            return ENGINEERING
        elseif (mod == 2) then
            return POCKET_FACTORY
        endif
    elseif (curHero == BREWMASTER) then
        if (skillUlti) then
            return STORM_EARTH_AND_FIRE
        elseif (mod == 0) then
            return BREATH_OF_FIRE
        elseif (mod == 1) then
            return DRUNKEN_HAZE
        elseif (mod == 2) then
            return DRUNKEN_BRAWLER
        endif
    elseif (curHero == FIRELORD) then
        if (skillUlti) then
            return VOLCANO
        elseif (mod == 0) then
            return INCINERATE
        elseif (mod == 1) then
            return SOUL_BURN
        elseif (mod == 2) then
            return SUMMON_LAVA_SPAWN
        endif
    endif
```

and

```
call SetBuildUpgr( 1000, UPG_CHEAP_EVOLUTION )
```

**Comment by tdauth** [comment]
Apparently The Allianced caused the crashes.

**Comment by tdauth** [comment]
The Alliance AI script had the Power Generator twice in it.

**Comment by tdauth** [comment]
Compare to https://github.com/tdauth/wowr/blob/b3195a497a2a65bd4abfb6bd44fbfef881899dc2/ai/TheAlliance.ai or the second after this one. It did not crash here. Now it is crashing again.

**Comment by tdauth** [comment]
Closed because of #1690.
